### PR TITLE
Add CrashingRunner for use in TestPipeline

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

CrashingRunner is a PipelineRunner that crashes on calls to run() with
an IllegalArgumentException. As a runner is currently required to
construct a Pipeline object, this allows removal of all Pipeline Runners
from the core SDK while retaining tests that depend only on the graph
construction behavior.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+class CrashingRunner extends PipelineRunner<PipelineResult>{
+
+  public static CrashingRunner fromOptions(PipelineOptions opts) {
+    return new CrashingRunner();
+  }
+
+  @Override
+  public PipelineResult run(Pipeline pipeline) {
+    throw new IllegalArgumentException(String.format("Cannot call #run(Pipeline) on an instance "
+            + "of %s. %s should only be used as the default to construct a Transform Hierarchy "
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java]: s/Transform Hierarchy/Pipeline/ I'd say. Also choose whether you want to use a capitalized class name or English prose, but avoid the mix. Here you have also "Pipeline Options" which probably should be PipelineOptions. If you want it to just be prose then "options" or "pipeline options".

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        getClass().getSimpleName(),
+        TestPipeline.class.getSimpleName(),
+        PipelineRunner.class.getSimpleName(),
+        TestPipeline.PROPERTY_BEAM_TEST_PIPELINE_OPTIONS));
+  }
+
+  private static class TestPipelineResult implements PipelineResult {
+    private TestPipelineResult() {
+      // Should never be instantiated by the enclosing class
+      throw new AssertionError(String.format("Forbidden to instantiate %s",
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java]: I think `checkState` or `UnsupportedOperationException`?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 3]
```diff
@@ -139,6 +141,15 @@ public void testMatcherSerializationDeserialization() {
     assertEquals(m2, newOpts.getOnSuccessMatcher());
   }
 
+  @Test
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: Part of the point, though, is that it does not fail if run() is not called, so I'd explicitly test that.

Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CrashingRunner.java]: Done.

**Comment by kennknowles** [comment]
LGTM

**Comment by tgroh** [comment]
Closed in dc98211

**Pull request closed without merging**